src/regdesk/printer_manager.js

```diff
@@ -0,0 +1,171 @@
+    if (dropdown.printer) {
+      // Spurious event dispatch?
+      return;
+    }
+
+    this.pairPrinter(dropdown);
+  }
+
+  /**
+   *
+   * @param {Object} o - Event Object.
+   * @param {USBDevice} o.device - The USB device to connect to as a printer.
+   * @returns
+   */
+  async handleConnectPrinter({ device }) {
+    // If we already have a tracked printer we shouldn't try to add it.
+    if (this.#dropdowns.some((d) => d.printer?.device === device)) {
+      return;
+    }
+
+    let printer = new this.#printerType(device);
+    try {
+      await printer.connect();
+    } catch (e) {
+      // TODO: Display something more useful.
+      console.error("Failed to connect to printer: " + e);
+    }
+
+    return printer;
+  }
+
+  /**
+   * Disconnect a printer from the page.
+   * @param {Object} o - Container object
+   * @param {USBDevice} o.device - The device to disconnect.
+   * @returns
+   */
+  async handleDisconnectPrinter({ device }) {
+    let dropdown = this.#dropdowns.filter(d => d.printer?.device === device);
+    if (!dropdown) {
+      return;
+    }
+
+    await dropdown.removePrinter();
+  }
+
+  /**
+   * Add a new printer to the page and connect to it.
+   * @param {PrinterDropdown} dropdown - The dropdown to add the printer to.
+   * @returns {LP2844} connected printer.
+   */
+  async pairPrinter(dropdown) {
+    var device;
+    try {
+      device = await navigator.usb.requestDevice({
+        filters: [
+          { vendorId: this.#printerType.usbVendorId }
+        ]
+      });
+    } catch (e) {
+      // TODO: One of the exceptions we can catch here is "user clicked cancel."
+      // Figure out how to separate that out as 'not an actual exception'.
+      console.error("Couldn't add new printer: " + e);
+      return undefined; // Nothing to return, can't proceed here.
+    }
+
+    let printer = await this.handleConnectPrinter({ device: device });
+    if (printer) {
+      this.assignPrinter(printer, dropdown);
+    }
+  }
+
+  /**
+   * Print a badge label to the adult printer.
+   * @param {LabelEpl} label
+   */
+  async printAdultLabel(label) {
+    if (!this.#adultDropdown.printer) {
+      // TODO: Make this more clear
+      console.error("Adult printer not present to print to!");
+    }
+
+    await this.#adultDropdown.printer.printLabel(label);
+  }
+
+  /**
+   * Print a badge label to the minor printer.
+   * @param {LabelEpl} label
+   */
+  async printMinorLabel(label) {
+    if (!this.#minorDropdown.printer) {
+      // TODO: Make this more clear
+      console.error("Minor printer not present to print to!");
+    }
+
+    await this.#minorDropdown.printer.printLabel(label);
+  }
+
+  /**
+   * Disconnect and reconnect to all printers. NOTE: May shuffle position of printers!
+   */
+  async refreshPrinters() {
+    this.#dropdowns.forEach(async d => {
+      await d.removePrinter();
+    })
+
+    navigator.usb.getDevices().then((devices) => {
```

Comment:
Missing an await (or return) statement on this promise, this async method behavior is undefined otherwise.